### PR TITLE
Feat: Add backend support for trip invites

### LINKED ISSUE #291: Create backend support for trip invites

### PR DESCRIPTION
**Pull request by g0g05arui**
## Summary

In this PR I introduce the necessary backend functions / model for the trip invites notification:

## Changes

- **Models:** Added the `TripInvite` model which has  `tripId`, `id`, `from` and `to` fields
- **Repositories:** I created the `TripInviteRepository` interface and the `TripInviteRepositoryFirebase` which implements this interface

## Checklist
- [x] This PR resolves #291 
- [x] Tests have been added for new functionality.
- [x] Documentation has been updated (if applicable).

##  Testing
- **Manual Testing:**
    - [ ] Tested on emulator / physical device.
- **Unit Tests**
    - [x] Added units tests for this

### CODE REVIEW

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+ * Used for displaying trip invites and notifications Appears as a parameter to the ${UserViewModel}
+ *
+ * @param db the firestore instance to use
+ * @param collectionPath the path to the collection of trip invites
+ * @constructor creates a new TripInviteRepositoryFirebase
+ * @property db the firestore instance to use
+ * @property collectionPath the path to the collection of trip invites
+ */
+class TripInviteRepositoryFirebase(private val db: FirebaseFirestore) : TripInviteRepository {
+  private val collectionPath = "friendRequests"
```
Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:18]: The collection path should be updated to something more appropriate, such as "tripInvites", to better reflect the purpose of the data it stores. While avoiding hardcoded strings is generally a good practice for maintainability, it's less critical in this backend context. That said, extracting the collection path into a constant can still improve clarity and make future updates easier.

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+      req: TripInvite,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    db.collection(collectionPath)
+        .document(req.id)
+        .set(req, SetOptions.merge())
+        .addOnSuccessListener { onSuccess() }
+        .addOnFailureListener { exception -> onFailure(exception) }
+  }
```
Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:61]: You are assuming that req is always valid, but it would be safer to include a null or empty check to handle potential issues. We've encountered bugs in the past caused by invalid input, which could arise from network errors, data corruption, or deserialization issues. Additionally, malformed or incomplete data from external sources (e.g., third-party integrations or API calls) could lead to unexpected behavior. Adding a validation step here would make the code more robust and resilient to these scenarios, reducing the likelihood of similar issues in the future.

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+      invite: String,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    db.collection(collectionPath)
+        .document(invite)
+        .delete()
+        .addOnSuccessListener { onSuccess() }
+        .addOnFailureListener { exception -> onFailure(exception) }
+  }
```
Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:102]: Same here

**Comment by BogdanCDev** [comment]
The implementation looks solid overall and effectively addresses the requirements. Thank you for ensuring the target test coverage is achieved—this adds significant value to the project’s reliability. Please consider the previously mentioned suggestions regarding validation and hardcoded strings.

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt (hunk 2 above)
Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:61]: > Additionally, malformed or incomplete data from external sources (e.g., third-party integrations or API calls) could lead to unexpected behavior.

Are we using any 3rd party integrations and API calls that I am not aware of? Also when this happens we have the `onFailureListner`. Seems like an AI-generated comment 
image

But I will update the code as you requested 😉

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt (hunk 1 above)
Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:18]: You are right, I forgot to update the collection name, I will rename it asap.

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:61]: You are partially correct that we have an onFailureListener, which handles Firebase-specific errors such as network issues, permission errors, or problems during the write operation. Nevertheless, it does not protect against issues that occur before the Firestore call, such as accessing req.id when req is null, which could throw a NullPointerException. This is why it's crucial to validate inputs before invoking Firestore operations.  If req is null, any attempt to access its properties will cause the app to crash before Firestore is even called. This will bypass the onFailureListener, making it ineffective in such cases.  Even if req is non-null, if req.id is null or blank, Firestore may attempt to create a document with an invalid ID or fail silently, leading to potential data integrity issues.

Comment (mirceagavrizi) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:61]: Hi @BogdanCDev I can't help but notice that this comment is also AI-generated. I think that it would be beneficial for this project, but also for yourself as a future software engineer, that you try to find the eventual mistakes/bugs on your own. Also, for @g0g05arui in this case, it seems like he is talking with a robot, not with his teammate. I think this applies for other PRs as well... 😉 
image

Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:61]: > If req is null, any attempt to access its properties will cause the app to crash before Firestore is even called.


As I mentioned before: This can never happen as the parameter doesnt use a `?` THIS ENSURES that null checking is not required. Even if I try to null check the editor said it is not needed. If you want I can refer you to the official kotlin documentation...

I will check that the id is notEmpty, but this is a different case.

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt (hunk 3 above)
Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt:102]: This is not the same. If the document doesn't exist, the onFailure event will be raised

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: Even though in pure kotlin code this is true, kotlin can be used can be used interchangeably with java, where null-safety is not enforced. This means that if the function is called from Java, null values could still be passed, bypassing Kotlin's guarantees. It might be the case in the future when we call the function with parameters given by deserialisation error from Firestore, interaction with API that might use java code, or endless other edge case examples that might occur. Even if it might seem redundant I personally prefer to code in a more defensive way and to make sure that the app cannot crush under any circumstances. Also, the overhead is really small and I don't really get why you are so bothered by such a change.

Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: While I still believe that a null check might be useful and might prevent future bugs, since you are so bothered by this change, which I somehow understand, since it does not pose an immediate issue,  I will accept this pr as the overall implementation is solid, but I personally believe that more defensive programming technics might be useful in the long run.

**Review by BogdanCDev — APPROVED** [review]
While I still believe that a null check might be useful and might prevent future bugs, since you are so bothered by this change, which I somehow understand, since it does not pose an immediate issue, I will accept this pr as the overall implementation is solid. Still, I personally believe that more defensive programming technics might be useful in the long run.

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt (hunk 2 above)
Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: No don't get me wrong, I am not bothered by the ask to change the code (which I did), just by the use of AI to generate really long comments that are not beneficial or even aplicable in some cases (this for example). This isn't just bothering me, the other teammates as well.

> Even though in pure kotlin code this is true, kotlin is fully interoperable with java, where null-safety is not enforced. This means that if the function is called from Java, null values could still be passed, bypassing Kotlin's guarantees. It might be the case in the future when we call the function with parameters given by deserialisation error from Firestore, interaction with API that might use java code, or endless other edge case examples that might occur.

We are not inter oping with java and this function is only called by kotlin code so this does not apply here

However, I have updated the code, even though the assumptions you made are flawed.

**Comment by g0g05arui** [comment]
Great, thanks for the review! Solid comments and feedback!

File: app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: I am sorry that you did not find my comment beneficial. Thank you for the suggestions and feedback. I will make sure to take it into account your suggestions in the future and try to make smaller comments, since I can understand that detailed explanations might be hard to follow. Regarding the use of AI, I must state the fact that I did not use it in the finding of any bug or suggestion I made.
Ps.: The same empty check should have been implemented also in deleteTripInvite, since you agreed that this was a beneficial feedback, but maybe you could take that into account in a future pr.

Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: I am sorry that you did not find my comment beneficial. Thank you for the suggestions and feedback. I will make sure to take it into account your suggestions in the future and try to make smaller comments, since I can understand that detailed explanations might be hard to follow. Regarding the use of AI, I must state the fact that I did not use it in the finding of any bug or suggestion I made.

Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: It's not about comments being too hard to follow or too long, but the AI leaves bad or even unnecessary suggestions (see the example above about 3rd party apis and null checking a non null enforced parameter) just for the sake of leaving a detailed reviewed / more comments.  Sometimes code is just straight forward and doesn't require updates. But we should end this conversation here and discuss this in the next stand up or in the next meeting with the coaches if you dont respect our opinions...

Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: You are assuming that req is always valid, but it would be safer to include a null or empty check to handle potential issues. We’ve encountered bugs in the past caused by invalid input, which could arise from network errors, data corruption, or deserialization issues. Additionally, malformed or incomplete data from external sources(e.g., third-party integrations or API calls) could lead to unexpected behavior. Adding a validation step here would make the coee more robust and resilient to these scenarios, reducing the likelihood of similar issues in the future.

Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/notifications/TripInviteRepositoryFirebase.kt]: Yes, find the whole conversation [here](https://github.com/swent-t16/Voyageur/pull/290/files/d8855fb0be99a8bcf960272e1be63df275b28a64#r1874999422)
